pkg/observable/channel/observable_test.go

```diff
@@ -0,0 +1,366 @@
+		// robustly (no flakiness); perhaps it has to do with the lack of some
+		// kind of guarantee about the receiver order on the consumer side.
+		//
+		// The following scenarios should generally pass but are flaky:
+		//
+		// {
+		// 	name:            "full non-buffered producer",
+		// 	producer:        fullBlockingProducer,
+		// 	inputs:          inputs[1:],
+		// 	expectedOutputs: inputs,
+		// 	setupFn: func(t test) {
+		// 		go func() {
+		// 			// blocking send
+		// 			t.producer <- &inputs[0]
+		// 		}()
+		// 	},
+		// },
+		// {
+		// 	name:            "full buffered len 1 producer",
+		// 	producer:        fullBufferedProducer,
+		// 	inputs:          inputs[1:],
+		// 	expectedOutputs: inputs,
+		// 	setupFn: func(t test) {
+		// 		// non-blocking send
+		// 		t.producer <- &inputs[0]
+		// 	},
+		// },
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setupFn != nil {
+				tt.setupFn(tt)
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			t.Cleanup(cancel)
+
+			obsvbl, producer := channel.NewObservable[*int](
+				channel.WithProducer(tt.producer),
+			)
+			require.NotNil(t, obsvbl)
+			require.NotNil(t, producer)
+			produce := produceWithDelay(producer, productionDelay)
+
+			// construct 3 distinct observers, each with its own channel
+			observers := make([]observable.Observer[*int], 1)
+			for i := range observers {
+				observers[i] = obsvbl.Subscribe(ctx)
+			}
+
+			group, ctx := errgroup.WithContext(ctx)
+
+			// ensure all obsvr channels are notified
+			for obsvrIdx, obsvr := range observers {
+				next := func(outputIndex int, output *int) error {
+					// obsvr channel should receive notified input
+					if !assert.Equalf(
+						t, tt.expectedOutputs[outputIndex],
+						*output,
+						"obsvr Idx: %d", obsvrIdx,
+					) {
+						return fmt.Errorf("unexpected output")
+					}
+					return nil
+				}
+
+				done := func(outputs []*int) error {
+					if !assert.Equalf(
+						t, len(tt.expectedOutputs),
+						len(outputs),
+						"obsvr addr: %p", obsvr,
+					) {
+						return fmt.Errorf("unexpected number of outputs")
+					}
+					return nil
+				}
+
+				// concurrently await notification or timeout to avoid blocking on
+				// empty and/or non-buffered producers.
+				group.Go(goNotifiedOrTimedOutFactory(obsvr, next, done, notifyTimeout))
+			}
+
+			// notify with test input
+			for _, input := range tt.inputs {
+				inputPtr := new(int)
+				*inputPtr = input
+
+				// simulating IO delay in sequential message production
+				produce(inputPtr)
+			}
+			cancel()
+
+			// wait for obsvbl to be notified or timeout
+			err := group.Wait()
+			require.NoError(t, err)
+
+			// unsubscribing should close obsvr channel(s)
+			for _, observer := range observers {
+				observer.Unsubscribe()
+
+				// must drain the channel first to ensure it is closed
+				err := testchannel.DrainChannel(observer.Ch())
+				require.NoError(t, err)
+			}
+		})
```

Comment:
Should we add a check that the observable is closed?

Reply:
Currently `Observable` doesn't have a "closed" state. Calling `#Close()` actually unsubscribes all observers; perhaps it should be renamed to `#UnsubscribeAll()`. To answer the spirit of your question. There is no public API to test whether the observable has any observers currently subscribed. I think we can (if we're not already) get coverage of that elsewhere, likely in  a smaller unit test.